Formal2024/Workshops/workshop5.lean

```diff
@@ -0,0 +1,402 @@
+-- is well-founded, i.e. it will always terminate. This is thanks to the theory of inductive
+-- types, which Lean is aware of.
+
+lemma add_zero (n : MyNat) : add n zero = n := rfl
+lemma add_succ (n m : MyNat) : add n (succ m) = succ (add n m) := rfl
+
+-- Using pattern matching or `racses`/`rintro` is not always enough to prove things about
+-- inductive types. Sometimes we need to use *induction*.
+lemma zero_add (n : MyNat) : add zero n = n := by
+  induction n with
+  | zero => rfl
+  | succ n ih => rw [add_succ, ih]
+
+-- There is also the `induction'` tactic, which is more powerful (see `generalizing`) and
+-- makes the diferent cases less explicit.
+lemma zero_add' (n : MyNat) : add zero n = n := by
+  induction' n with n ih
+  · rfl
+  · rw [add_succ, ih]
+
+-- What happens if instead of `induction'` we use `rcases`?
+
+-- Lean is able to do induction on any inductive type. Every time we define an inductive type,
+-- Lean generates some induction principles for us. These are called `rec` and `recOn`, and they
+-- are what the `induction` tactic uses under the hood.
+#check MyNat.rec
+-- MyNat.rec.{u}
+--   {motive : MyNat → Sort u}
+--   (zero : motive zero)
+--   (succ : (n : MyNat) → motive n → motive (succ n))
+--   (t : MyNat)
+-- : motive t
+#check MyNat.recOn
+
+/-
+  # Exercises
+-/
+
+-- 1. Prove that addition is associative. (Can you do it in 3 lines?)
+
+lemma add_assoc (n m k : MyNat) : add (add n m) k = add n (add m k) := by
+  sorry
+
+-- Hint: we defined what `add n (succ m)`, with the succesor on the right.
+-- Does it help to induct on the rightmost argument?
+
+-- 2. Define multiplication in terms of addition.
+
+def mul : MyNat → MyNat → MyNat
+  | n, zero => zero
+  | n, succ m => sorry
+
+lemma mul_zero (n : MyNat) : mul n zero = zero := rfl
+lemma mul_succ (n m : MyNat) : mul n (succ m) = add (mul n m) n := sorry
+
+-- Now prove that multiplying by zero on the left gives zero.
+
+lemma zero_mul (n : MyNat) : mul zero n = zero := by
+  sorry
+
+end MyNat
+end
+
+section
+
+-- The constructors of an inductive type can take more than one argument.
+inductive MyExists {α : Type} (p : α → Prop) : Prop where
+  | intro (w : α) (h : p w) : MyExists p
+
+example : MyExists (fun x : ℕ ↦ x > 0) := MyExists.intro 1 (by norm_num)
+
+-- When we have a inductive type with only one constructor, we can use the *anonymous constructor*.
+example : MyExists (fun x : ℕ ↦ x > 0) := ⟨1, by norm_num⟩
+
+example (h : MyExists (fun x : ℕ ↦ x < 0)) : False := by
+  rcases h with ⟨n, hn⟩
+  norm_num at hn
+
+-- Inductive types with only one constructor are called *structures*. We will learn more
+-- about these next week!
+
+-- If you are curious, you can `ctrl/cmd + click` on these to see how they are defined in Lean as
+-- inductive types.
+#check And
+#check Or
+#check Exists
+#check Iff
+#check ℕ
+#check ℤ
+#check Prod
+#check Unit
+#check Empty
+
```

Comment:
My only concern is that most of these are structures. We say that a structure is an inductive type with only one constructor, but without knowing the syntax if they Ctr click for example Prod they'll see something that looks like an inductive type with two constructors.

Reply:
That's fair. I'll add a brief comment saying how the structure syntax works, hopefully not spoiling too much for next week.